### ISSUE TITLE
Umbraco 9.5.1 - VS publish deploy issue

### ISSUE THREAD

**Issue by thomashdk** [umbraco/Umbraco-CMS#12572]
### Which *exact* Umbraco version are you using? For example: 9.0.1 - don't just write v9

9.5.1

### Bug summary

After upgrading to 9.5.1, when publish deploy from visual studio the publish skips *.html files from the ./App_Plugins folder.
Works fine in 9.4.3 - Stopped working in 9.5.0
### Specifics

Html files in App_Plugins folder are missing after a Folder Publish deploy from Visual Studio

### Steps to reproduce

- Install Umbraco 9.5.1
- Create a App_Plugins folder
- Create a plugin folder.
- Create a html files.

-Publish project to a folder
- Html file in plugin are missing in the deploy folder

### Expected result / actual result

All included files should be copied to deploy folder when included in project.

---
_This item has been added to our backlog [AB#25526](https://dev.azure.com/umbraco/243e7927-03b2-44e2-908f-d4ac7ea5daaa/_workitems/edit/25526)_

**Comment by Ambertvu** [comment]
I'm having the exact same thing since the upgrade. 
Only thing is I'm not publishing from visual studio but from an Azure Pipeline (and it creates a docker container) so couldn't easily check if it was a file capitalization issue or if something went wrong during the publish. But this at least saves time debugging tonight ;-)

I double checked my commits, and the only thing changed is the version number in the .proj files. so not entirely sure why these files aren't published anymore.

**Comment by thomashdk** [comment]
> But this at least saves time debugging tonight ;-)

I spend way to much time finding out what the issue could be.. :P

**Comment by Ambertvu** [comment]
This also still happens in V10.0.0

**Comment by Ambertvu** [comment]
I've been testing this a bit (cleaned up my `.proj` so it contained nothing except the default. Re-added my custom Plugin. 
It was added like this:
```
      <Content Include="App_Plugins\Top10picker\top10picker.html"/>
    </ItemGroup>
	<ItemGroup>
		<None Include="App_Plugins\Top10picker\package.manifest" />
		<None Include="App_Plugins\Top10picker\top10picker.controllers.js" />
		<None Include="App_Plugins\Top10picker\top10picker.cshtml" />
	</ItemGroup>
```
After a republish the .html file was still missing.

Changed the properties for the .html file to Copy Always:
```    <ItemGroup>
      <Content Include="App_Plugins\Top10picker\top10picker.html">
        <CopyToOutputDirectory>Always</CopyToOutputDirectory>
      </Content>
    </ItemGroup>
	<ItemGroup>
		<None Include="App_Plugins\Top10picker\package.manifest" />
		<None Include="App_Plugins\Top10picker\top10picker.controllers.js" />
		<None Include="App_Plugins\Top10picker\top10picker.cshtml" />
	</ItemGroup>```

This fixed it for me as a workaround. But this shouldn't be default behaviour.

I had the exact same issue with /config/grid.editors.config.js, only when specifically saying to  Copy Always it in the properties it would be fixed.

**Comment by Zeegaan** [comment]
Sorry for the late answer! I can definitely reproduce this, and IIRC we've had a problem like this in the past (I think I even tried to fix it) 🤣 Interesting note here is that if my file is directly in the App_Plugins folder it gets published fine: 
![image](https://user-images.githubusercontent.com/70372949/209803994-569ea3a0-940e-4c90-be07-ef5772dc9597.png)
But if its in another folder it does not get published 🤔

**Comment by jaandrews** [comment]
@eliranrefael One way I've used to work around this issue is to move all of the files into a razor class library. This bundles the files into the package dll and so avoids this problem. This is what umbraco forms does now, which is why it's files no longer show under the app_plugins folder. Some guidance on how to work with this can be found [here](https://dev.to/kevinjump/umbraco-10-razor-class-library-packages-pt1-3nfa).

This won't help though if the package isn't set up for this though, in which case you will need to tinker with the project files as suggested above. A quick hack that may help would be to replace all references that are under app_plugins with a single
```
<ItemGroup>
      <Content Include="App_Plugins\**" CopyToOutputDirectory="Always" />
</ItemGroup>
```
This should make the files deploy, but keep in mind that this will prevent any cs files from compiling, as those need a "Compile" tag and not a "Content" tag to compile. Can still be made to work, but it would need to target specific directories/files instead of the whole app_plugins folder, like how it's done in the earlier [example](https://github.com/umbraco/Umbraco-CMS/issues/12572#issuecomment-1163415585).

**Comment by eliranrefael** [comment]
@jaandrews hey,
in this stage i would be satisfied in just including the folder as u suggested. the problem is that the solution you suggested doesnt work. for demonstration, delete the folder from source before publishing. it seems like the folder is created only after publish is ended, like in post build actions.

**Comment by jaandrews** [comment]
@eliranrefael Did you delete all other references to the files in the app_plugins folder in the project file aside from the catch all reference? Does it work as expected if you run dotnet publish on your local machine? This did copy the files when I did that and was the workaround suggested when I contacted umbraco support about this issue. Could be a problem with your deployment pipeline if that works on your end.

**Comment by bergmania** [comment]
I have spend some time looking into this.

From what I can see, then it works as expected out of box, BUT (!), when I add the files using by IDE, then it automatically adds the following to my project file (`.csproj`)
```xml
<ItemGroup>
  <Content Include="App_Plugins\FolderNameFileName.ext" />
</ItemGroup>
```
This states that the file should not be copied, because it do not set the `CopyToPublishDirectory`.

If I wanna see all files from `App_Plugins` in my IDE, I can add this:
```xml
<ItemGroup>
  <Content Include="App_Plugins\**">
    <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
    <CopyToPublishDirectory>PreserveNewest</CopyToPublishDirectory>
  </Content>
</ItemGroup>
 ```
 Then it works again and all files seems to be copied to my publish folder.
 
Personally, I do not think it make sense to add this automatically to our templates, as many packages still write content to `App_Plugins`, instead of using Razor Class Libraries. Files from packages should IMO not be shown in my project overview.

**Comment by bergmania** [comment]
I'll close this issue as it works as expected in v10+

I'm sorry you experienced this on a minor update :(

**Comment by lassefredslund** [comment]
Just FYI, here is my best attempt at an explanation.

Clients sometimes experience that they have a project that works locally, but when pushing it to Cloud (or any other hosting environment) some files go missing from the App_Plugins folder.
This happens because IDEs like Visual Studio and Rider (but not VS Code) try to assist and play intelligent with files that are not added to the wwwroot-folder, but it seems to confuse many developers in this specific case.
In short, there are two ways to avoid this:

1. Place your App_Plugins folder below the wwwroot folder, as all files will be copied during publishing. If you choose this option, please be sure that you do not have any license files or files that should not be directly accessible through a URL.
2. Place your App_Plugins folder outside the wwwroot folder and let the IDE know about the specific files you want to copy during publishing. By default the IDE includes files in the project through the project files like this:
![image](https://user-images.githubusercontent.com/101121881/216331132-1fd9f804-b8ee-4de7-b664-7eee55ebf24e.png)

However, to ensure that files are also included when publishing, you need to change Include to Update like this or explicitly specify what files to copy on publish. Please ensure not to include e.g. license files and other files that should not be directly accessible through the URL:
![image](https://user-images.githubusercontent.com/101121881/216330701-dc0bd190-6f19-4dcb-9521-cfc46a18aa08.png)

**History**
The description above is 100% standard Visual Studio/Rider experience, so one might argue that no developers should be surprised about this. However, we need to recall the history here, as it hasn’t always been this way. Up until version 9.4, we actually did some “magic” in the project-file, so that App_Plugins files automatically were included, however, with V9.5 this logic was changed. We still copy over files that are located in App_Plugins folder (even if this folder is located outside the wwwroot folder) but if the files are listed in the project file, we only publish files that are set to “update”.
